File: js/examples/config_mlp_fit.js
'use strict'

// Run from repo root:
//   POLY_CORE=native node js/examples/config_mlp_fit.js

const { create } = require('../src')

async function main() {
  const pg = await create({ core: process.env.POLY_CORE || 'auto' })
  const inst = pg.Instance.mlp({
    layers: [2, 4, 1],
    activation: 'relu',
    bias: true,
    loss: 'mse',
    batch_size: 1,
    seed: 42
  })

  const losses = inst.fit({
    x: new Float32Array([1, 2]),
    y: new Float32Array([4])
  }, { epochs: 12, optimizer: 'sgd', lr: 0.03 })

  console.log('loss', losses[0], '->', losses[losses.length - 1])
  console.log('output', Array.from(inst.forward({ x: new Float32Array([1, 2]) }).output))

  await pg.dispose()
}

main().catch(err => {
  console.error(err)
  process.exitCode = 1
})

File: py/examples/config_mlp_fit.py
"""Config-based MLP instance training.

This is the family-template path used by built-in/imported models.

Run from repo root:
  PYTHONPATH=py POLYGRAD_LIB=build/libpolygrad.so python py/examples/config_mlp_fit.py
"""

import numpy as np

from polygrad.instance import Instance


inst = Instance.mlp({
    "layers": [2, 4, 1],
    "activation": "relu",
    "bias": True,
    "loss": "mse",
    "batch_size": 1,
    "seed": 42,
})

history = inst.fit(
    {"x": np.array([[1.0, 2.0]], dtype=np.float32),
     "y": np.array([[4.0]], dtype=np.float32)},
    epochs=12,
    optimizer="sgd",
    lr=0.03,
)

print("loss", history[0], "->", history[-1])
print("output", inst.forward(x=np.array([[1.0, 2.0]], dtype=np.float32))["output"])
